File: tests/algorithms/x/OHP/strings/is_spain_national_id.php
<?php
error_reporting(E_ALL & ~E_DEPRECATED);
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _panic($msg) {
    fwrite(STDERR, strval($msg));
    exit(1);
}
$__start_mem = memory_get_usage();
$__start = _now();
  $DIGITS = '0123456789';
  $UPPER = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
  $LOWER = 'abcdefghijklmnopqrstuvwxyz';
  $LOOKUP_LETTERS = 'TRWAGMYFPDXBNJZSQVHLCKE';
  $ERROR_MSG = 'Input must be a string of 8 numbers plus letter';
  function to_upper($s) {
  global $DIGITS, $ERROR_MSG, $LOOKUP_LETTERS, $LOWER, $UPPER;
  $res = '';
  $i = 0;
  while ($i < strlen($s)) {
  $ch = substr($s, $i, $i + 1 - $i);
  $j = 0;
  $converted = $ch;
  while ($j < strlen($LOWER)) {
  if (substr($LOWER, $j, $j + 1 - $j) == $ch) {
  $converted = substr($UPPER, $j, $j + 1 - $j);
  break;
}
  $j = $j + 1;
};
  $res = $res . $converted;
  $i = $i + 1;
};
  return $res;
};
  function is_digit($ch) {
  global $DIGITS, $ERROR_MSG, $LOOKUP_LETTERS, $LOWER, $UPPER;
  $i = 0;
  while ($i < strlen($DIGITS)) {
  if (substr($DIGITS, $i, $i + 1 - $i) == $ch) {
  return true;
}
  $i = $i + 1;
};
  return false;
};
  function clean_id($spanish_id) {
  global $DIGITS, $ERROR_MSG, $LOOKUP_LETTERS, $LOWER, $UPPER;
  $upper_id = to_upper($spanish_id);
  $cleaned = '';
  $i = 0;
  while ($i < strlen($upper_id)) {
  $ch = substr($upper_id, $i, $i + 1 - $i);
  if ($ch != '-') {
  $cleaned = $cleaned . $ch;
}
  $i = $i + 1;
};
  return $cleaned;
};
  function is_spain_national_id($spanish_id) {
  global $DIGITS, $ERROR_MSG, $LOOKUP_LETTERS, $LOWER, $UPPER;
  $sid = clean_id($spanish_id);
  if (strlen($sid) != 9) {
  _panic($ERROR_MSG);
}
  $i = 0;
  while ($i < 8) {
  if (!is_digit(substr($sid, $i, $i + 1 - $i))) {
  _panic($ERROR_MSG);
}
  $i = $i + 1;
};
  $number = intval(substr($sid, 0, 8));
  $letter = substr($sid, 8, 8 + 1 - 8);
  if (is_digit($letter)) {
  _panic($ERROR_MSG);
}
  $expected = substr($LOOKUP_LETTERS, $number % 23, $number % 23 + 1 - $number % 23);
  return $letter == $expected;
};
  function main() {
  global $DIGITS, $ERROR_MSG, $LOOKUP_LETTERS, $LOWER, $UPPER;
  echo rtrim(json_encode(is_spain_national_id('12345678Z'), 1344)), PHP_EOL;
  echo rtrim(json_encode(is_spain_national_id('12345678z'), 1344)), PHP_EOL;
  echo rtrim(json_encode(is_spain_national_id('12345678x'), 1344)), PHP_EOL;
  echo rtrim(json_encode(is_spain_national_id('12345678I'), 1344)), PHP_EOL;
  echo rtrim(json_encode(is_spain_national_id('12345678-Z'), 1344)), PHP_EOL;
};
  main();
$__end = _now();
$__end_mem = memory_get_peak_usage(true);
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;

File: tests/algorithms/transpiler/FS/strings/is_spain_national_id.fs
// Generated 2025-08-11 15:32 +0700

exception Break
exception Continue

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _substring (s:string) (start:int) (finish:int) =
    let len = String.length s
    let mutable st = if start < 0 then len + start else start
    let mutable en = if finish < 0 then len + finish else finish
    if st < 0 then st <- 0
    if st > len then st <- len
    if en > len then en <- len
    if st > en then st <- en
    s.Substring(st, en - st)

let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let DIGITS: string = "0123456789"
let UPPER: string = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
let LOWER: string = "abcdefghijklmnopqrstuvwxyz"
let LOOKUP_LETTERS: string = "TRWAGMYFPDXBNJZSQVHLCKE"
let ERROR_MSG: string = "Input must be a string of 8 numbers plus letter"
let rec to_upper (s: string) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable s = s
    try
        let mutable res: string = ""
        let mutable i: int = 0
        try
            while i < (String.length (s)) do
                try
                    let ch: string = string (s.[i])
                    let mutable j: int = 0
                    let mutable converted: string = ch
                    try
                        while j < (String.length (LOWER)) do
                            try
                                if (string (LOWER.[j])) = ch then
                                    converted <- string (UPPER.[j])
                                    raise Break
                                j <- j + 1
                            with
                            | Continue -> ()
                            | Break -> raise Break
                    with
                    | Break -> ()
                    | Continue -> ()
                    res <- res + converted
                    i <- i + 1
                with
                | Continue -> ()
                | Break -> raise Break
        with
        | Break -> ()
        | Continue -> ()
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
and is_digit (ch: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable ch = ch
    try
        let mutable i: int = 0
        while i < (String.length (DIGITS)) do
            if (string (DIGITS.[i])) = ch then
                __ret <- true
                raise Return
            i <- i + 1
        __ret <- false
        raise Return
        __ret
    with
        | Return -> __ret
and clean_id (spanish_id: string) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable spanish_id = spanish_id
    try
        let upper_id: string = to_upper (spanish_id)
        let mutable cleaned: string = ""
        let mutable i: int = 0
        while i < (String.length (upper_id)) do
            let ch: string = string (upper_id.[i])
            if ch <> "-" then
                cleaned <- cleaned + ch
            i <- i + 1
        __ret <- cleaned
        raise Return
        __ret
    with
        | Return -> __ret
and is_spain_national_id (spanish_id: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable spanish_id = spanish_id
    try
        let sid: string = clean_id (spanish_id)
        if (String.length (sid)) <> 9 then
            failwith (ERROR_MSG)
        let mutable i: int = 0
        while i < 8 do
            if not (is_digit (string (sid.[i]))) then
                failwith (ERROR_MSG)
            i <- i + 1
        let number: int = int (_substring sid 0 8)
        let letter: string = string (sid.[8])
        if is_digit (letter) then
            failwith (ERROR_MSG)
        let expected: string = string (LOOKUP_LETTERS.[((number % 23 + 23) % 23)])
        __ret <- letter = expected
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        printfn "%b" (is_spain_national_id ("12345678Z"))
        printfn "%b" (is_spain_national_id ("12345678z"))
        printfn "%b" (is_spain_national_id ("12345678x"))
        printfn "%b" (is_spain_national_id ("12345678I"))
        printfn "%b" (is_spain_national_id ("12345678-Z"))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()
